File: ios/FridgeSnap/Media/ImagePipeline.swift
import UIKit

// Client-side compression before upload (spec section 4.1): max 1568px on
// the long edge, JPEG quality 0.8. Keeps vision quality high and image
// tokens low.
enum ImagePipeline {
    static let maxLongEdge: CGFloat = 1568
    static let jpegQuality: CGFloat = 0.8

    static func targetSize(for size: CGSize, maxLongEdge: CGFloat = ImagePipeline.maxLongEdge) -> CGSize {
        let longEdge = max(size.width, size.height)
        guard longEdge > maxLongEdge, longEdge > 0 else { return size }
        let scale = maxLongEdge / longEdge
        return CGSize(width: (size.width * scale).rounded(), height: (size.height * scale).rounded())
    }

    static func compress(_ image: UIImage) -> Data? {
        let target = targetSize(for: image.size)
        if target == image.size {
            return image.jpegData(compressionQuality: jpegQuality)
        }
        let format = UIGraphicsImageRendererFormat()
        format.scale = 1
        let renderer = UIGraphicsImageRenderer(size: target, format: format)
        let resized = renderer.image { _ in
            image.draw(in: CGRect(origin: .zero, size: target))
        }
        return resized.jpegData(compressionQuality: jpegQuality)
    }
}

File: app/src/core/imagePipeline.ts
// Client-side compression math before upload (spec section 4.1): max 1568px on
// the long edge, JPEG quality 0.8. Ported from Swift ImagePipeline.swift. Only
// the pure targetSize math lives here so it unit-tests without any native
// module; the actual pixel resize wraps expo-image-manipulator in the media
// layer (see media/compressImage.ts, added in the UI phase).

export const MAX_LONG_EDGE = 1568;
export const JPEG_QUALITY = 0.8;

export interface Size {
  width: number;
  height: number;
}

export const targetSize = (
  size: Size,
  maxLongEdge: number = MAX_LONG_EDGE,
): Size => {
  const longEdge = Math.max(size.width, size.height);
  if (longEdge <= maxLongEdge || longEdge <= 0) return size;
  const scale = maxLongEdge / longEdge;
  return {
    width: Math.round(size.width * scale),
    height: Math.round(size.height * scale),
  };
};
